sdk/core/System.ClientModel/gen/AbstractTypeWithoutProxyAnalyzer.cs

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace System.ClientModel.SourceGeneration;
+
+/// <summary>
+/// Analyzer that ensures abstract types used as buildable models have a PersistableModelProxy defined.
+/// </summary>
+[DiagnosticAnalyzer(LanguageNames.CSharp)]
+public sealed class AbstractTypeWithoutProxyAnalyzer : DiagnosticAnalyzer
+{
+    /// <inheritdoc/>
+    public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics =>
+        ImmutableArray.Create(ModelReaderWriterContextGenerator.DiagnosticDescriptors.AbstractTypeWithoutProxy);
+
+    /// <inheritdoc/>
+    public override void Initialize(AnalysisContext context)
+    {
+        context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+        context.EnableConcurrentExecution();
+        context.RegisterCompilationStartAction(StartAction);
+    }
+
+    private static void StartAction(CompilationStartAnalysisContext context)
+    {
+        var buildableAttrType = context.Compilation.GetTypeByMetadataName("System.ClientModel.Primitives.ModelReaderWriterBuildableAttribute");
+        var proxyAttrType = context.Compilation.GetTypeByMetadataName("System.ClientModel.Primitives.PersistableModelProxyAttribute");
+        if (buildableAttrType == null)
+            return;
+
+        context.RegisterSymbolAction(symbolContext =>
+        {
+            var namedType = (INamedTypeSymbol)symbolContext.Symbol;
+            if (namedType.TypeKind != TypeKind.Class)
+                return;
+
+            // Only look at context classes (those inheriting from ModelReaderWriterContext)
+            var contextBaseType = symbolContext.Compilation.GetTypeByMetadataName("System.ClientModel.Primitives.ModelReaderWriterContext");
+            if (contextBaseType == null)
+                return;
```

Comment:
Prefer to fetch this in a `CompilationStartAction`, and then only register the symbol action if the type exists.